File: LeetCode/letter_combinations_of_a_phone_number.py
#!/usr/bin/env python
#coding: utf-8

class Solution:
    def _com(self, digits, d):
        if not digits: return ['']
        x = d[digits[0]]
        y = self._com(digits[1:], d)
        r = []
        for p in x:
            for q in y:
                r.append(p + q)
        return r

    # @return a list of strings, [s1, s2]
    def letterCombinations(self, digits):
        d ={'2':'abc', '3':'def', '4':'ghi',
            '5':'jkl', '6':'mno', '7':'pqrs',
            '8':'tuv', '9':'wxyz'}
        return self._com(digits, d)

if __name__ == '__main__':
    s = Solution()
    print(s.letterCombinations('23'))

File: LeetCode/Java/Solution/src/ios/ac/cn/LetterCombinationsofaPhoneNumber.java
package ios.ac.cn;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LetterCombinationsofaPhoneNumber {
	private static Map<Character, String> map = new HashMap<>();
	static {
		map.put('2', "abc");
		map.put('3', "def");
		map.put('4', "ghi");
		map.put('5', "jkl");
		map.put('6', "mno");
		map.put('7', "pqrs");
		map.put('8', "tuv");
		map.put('9', "wxyz");
	}

	public List<String> letterCombinations(String digits) {
		List<String> result = new ArrayList<>();
		if (digits != null && digits.length() > 0) {
			List<String> subResult = new ArrayList<>();
			if (digits.length() <= 1) {
				subResult.add("");
			} else {
				subResult = letterCombinations(digits.substring(1));
			}

			String prefix = map.get(digits.charAt(0));

			for (int i = 0; i < prefix.length(); i++) {
				for (String e : subResult) {
					StringBuilder builder = new StringBuilder();
					builder.append(prefix.charAt(i));
					builder.append(e);
					result.add(builder.toString());
				}
			}
		} else {
			result.add("");
		}
		return result;
	}

	public static void main(String[] args) {
		for (String e : new LetterCombinationsofaPhoneNumber()
				.letterCombinations("23")) {
			System.out.println(e);
		}
	}
}
